File: java/CountPrimes.java
/*
 * Filename: CountPrimes.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 25 Jun 2015 02:34:37 PM EDT
 */

public class Solution {
	public int countPrimes(int n) {
		if (n <= 2) {
			return 0;
		}

		boolean[] primes = new boolean[n];
		for (int i = 2; i < n; i++) {
			primes[i] = true;
		}

		for (int i = 2; i <= Math.sqrt(n-1); i++){
			if (primes[i]) {
				for (int j = 2*i; j < n; j += i) {
					primes[j] = false;
				}
			}
		}

		int count = 0;
		for (int i = 2; i < n; i++) {
			if (primes[i]) {
				count++;
			}
		}

		return count;
	}
}

File: cpp/count_primes.cpp
/*
 * Filename: count_primes.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 16 Nov 2015 11:26:23 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution {
	public:
		int countPrimes(int n) {
			if (n <= 2)
			{
				return 0;
			}

			int count = 0;
			bool is_composite[n];
			for (int i = 0; i < n; i++)
			{
				is_composite[i] = false;
			}

			for (int i = 2; i*i < n; i++)
			{
				if (!is_composite[i])
				{
					for (int j = i*i; j < n; j += i)
					{
						is_composite[j] = true;
					}
				}
			}

			for (int i = 2; i < n; i++)
			{
				if (!is_composite[i])
				{
					count++;
				}
			}

			return count;
		}
};
